### DISCUSSION TITLE
Dataloader with many to one relationship

### DISCUSSION THREAD

**Discussion by BengtHagemeister** [hayes/pothos#275] [Q&A] [answered]
Hi,

I cannot figure out how to use the dataloader with a many relationship. For example:

I have a user with many comments. The user_id is stored with the comments and I want to fetch all users with all their comments. In order to not have a n+1 problem I would like to use a dataloader.

```ts

type User = {
  id: number
  name: string
}

type Comment = {
  id: number;
  user_id: number;
  message: string
}

const UserObject = builder.objectRef<User>("User")
const CommentObject = builder.objectRef<Comment>("Comment")

const CommentLoadable =  builder.loadableObject(CommentObject, {
  load: () => ({................}),
  fields: (t) => ({
    id: t.exposeID("id"),
    message: t.exposeString("message")
  })
})

UserObject.implement({
  fields: (t) => ({
    id: t.exposeID("id"),
    comments: t.field({
      type: [CommentLoadable],
      resolve: (user) => user.id
    })
  })
})
```

When using a standard graphql/dataloader I would make a CommentDataloaderByUserId and load a group of result by user_id. But the `load`function only accepts an array of comments and not a double array.

Any help would be highly appreciated! Thanks.

**Comment by hayes** [comment]
Unfortunately this is something that is currently missing from the dataloader plugin.  Definitely something I should add support for, but for now, I think you just need to manually use a dataloader. 

It shouldn't be too hard to add a new field builder method for many-to-one relationships.  I'll probably add a `t.loadableList` method that lets the resovler return a single id, and expects the load to return a list of lists given a list of ids. 

I personally have not run into this yet, because most of my 1-to-many relationships could not be batch loaded, and 1-to-many relationships generally aren't re-used.  Definitely just an oversight on my part.

**Reply by BengtHagemeister** [reply]
I see. Thanks for the quick reply!

**Comment by hayes** [comment, marked as answer]
Added support for this in the latest version: https://www.pothos-graphql.dev/docs/plugins/dataloader#loadablelist-fields-for-one-to-many-relations

**Reply by BengtHagemeister** [reply]
Oh, thanks! That was very quick.

**Reply by hayes** [reply]
@babsonmatt This should work for your use case.  `resolve` would return the user id, then the `load` method receives a list of user ids (batched).  You then would return a array of arrays.  If you mostly care about caching and not batching, it should be as simple as:

```ts
comments: t.loadableList({
  type: Comment,
  resolve: user => user.id,
  load: ids => Promise.all(ids.map(id => getCommentsByUserId(id))
});
```

**Reply by babsonmatt** [reply]
Thanks a lot @hayes, I realized what I was doing wrong minutes after making the post so I deleted but thanks a lot for taking the time to reply and provide the solution!
